### PR TITLE
XML strings for the documentation should live outside of the src code, in xml files. 

### LINKED ISSUE #477: Additional documentation for the ML.Net types and components should live in xml docs
Through PR #455 we introduced XML strings in the codebase, that flow to the generated CSharpApi, to better document the trainers of the Microsoft.ML.Trainers namespace. 

The final destination for those XML strings should be xml files, in the 'docs' folder, parallel to the src folder. Those documents should be referenced in the CSharp api, and through the code by xml <includes> tags.

### PR DESCRIPTION
**Pull request by sfilipi**
Resolves #477  by moving the strings with the XML remarks, examples etc into a separate XML file. 
Now the actual classes and their C# Api counterparts share the path to the XML node that contains the documentation and the respective example. 

On this PR, there are only two examples that are separate from the rest of the descriptive xml: the LogisticRegressionBinaryClassifier and LogisticRegressionClassifier, to give an idea of the infrastructure. 

The rest of the examples are coming in the next PR, together with the rest of the documentation.

### CODE REVIEW

**Comment by TomFinley** [comment]
Thanks for doing this @sfilipi ! So what are best practices here? I might have expected that any XML include files would be in the same directory or some subdirectory as the project with the relevant code, as opposed to some other location way off in the hinterlands not even in the "source" files.

Secondary reason for my preference is that it might be nice if `docs` remain something human browsable and comprehensible.. right now it is that since they're just markdown files (which are human readable). This stuff, despite being "documentation" of a sort, should probably be closer to the source.

**Comment by TomFinley** [comment]
That is, I'd expect the FastTree XML docs to be somewhere in the FastTree project, LightGBM in lightGBM project, etc... rather than having master files where ***all*** documentation lives. I'm not sure I consider these massive files maintainable.

---
In reply to: [403604163](https://github.com/dotnet/machinelearning/pull/510#issuecomment-403604163) [](ancestors = 403604163)

**Comment by sfilipi** [comment]
Was on the fence between single file, per component type, and per-project files. 
I started with per-project, hence the PCA file. I'll update and break down, since the votes so far are 1.5 - 0.5 in favor of per-project files. 

---
In reply to: [403605251](https://github.com/dotnet/machinelearning/pull/510#issuecomment-403605251) [](ancestors = 403605251,403604163)

**Comment by TomFinley** [comment]
Hmmmm. :) Sorry what votes?

---
In reply to: [403606882](https://github.com/dotnet/machinelearning/pull/510#issuecomment-403606882) [](ancestors = 403606882,403605251,403604163)

**Comment by sfilipi** [comment]
The OSX failures are unrelated to my changes. They fail to retrieve some of the nuget dependancies.

File: src/Microsoft.ML.FastTree/FastTreeArguments.cs [hunk 1]
```diff
@@ -20,6 +20,7 @@ public interface IFastTreeTrainerFactory : IComponentFactory<ITrainer>
     {
     }
 
+    /// <include file='./XMLDoc.xml' path='docs/members/member[@name="FastTree"]/*' />
```
Comment (TomFinley) [inline src/Microsoft.ML.FastTree/FastTreeArguments.cs]: >XMLDoc [](start = 25, length = 6)

I might be being a little OCD, but the redundancy of the name `XMLDoc.xml` just drives me absolutely bonkers. Maybe just plain old `doc.xml` (lower case `doc`)?

File: src/Microsoft.ML.FastTree/XMLDoc.xml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+<docs>
+  <members>
+
+    <member name="FastTree">
+      <summary>
+        Trains gradient boosted decision trees to the LambdaRank quasi-gradient. 
+      </summary>
+      <remarks>
+        <para>
+          FastTrees is an efficient implementation of the <a href='https://arxiv.org/abs/1505.01866'>MART</a> gradient boosting algorithm.
```
Comment (TomFinley) [inline src/Microsoft.ML.FastTree/XMLDoc.xml]: >FastTrees [](start = 10, length = 9)

FastTree, not FastTrees.

File: src/Microsoft.ML.Core/EntryPoints/ModuleArgs.cs [hunk 3]
```diff
@@ -529,9 +529,9 @@ public sealed class EntryPointAttribute : Attribute
             public string ShortName { get; set; }
 
             /// <summary>
-            /// Remarks on the Entry Point, for more extensive XML documentation on the C#API
+            /// The path to the XML documentation on the C#API component
```
Comment (shauheen) [inline src/Microsoft.ML.Core/EntryPoints/ModuleArgs.cs]: >C#API [](start = 57, length = 5)

C# API?

File: src/Microsoft.ML.FastTree/doc.xml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+<docs>
+  <members>
+
+    <member name="FastTree">
+      <summary>
+        Trains gradient boosted decision trees to the LambdaRank quasi-gradient. 
+      </summary>
+      <remarks>
+        <para>
+          FastTree is an efficient implementation of the <a href='https://arxiv.org/abs/1505.01866'>MART</a> gradient boosting algorithm.
```
Comment (shauheen) [inline src/Microsoft.ML.FastTree/doc.xml]: >MART [](start = 100, length = 4)

This is a bit confusing, We say MART but the paper is an improved version DART. @tfinley@gmail.com do you know what is the correct one?

Comment (TomFinley) [reply src/Microsoft.ML.FastTree/doc.xml]: If we use the acronym "DART" I think we need to have a link to that paper, and also explain what is meant by that. I think it's fine to just say MART honestly. The DART capabilities could perhaps be mentioned.

**Pull request merged**